File: src/services/pathfinding.ts
import { Edge } from "./edges";

interface PriorityQueue {
  item: string;
  priority: number;
}

export class HeapPriorityQueue {
  queue: PriorityQueue[];

  constructor() {
    this.queue = [];
  }

  getCount() {
    return this.queue.length;
  }

  enqueue(item: string, priority: number) {
    this.queue.push({
      item,
      priority,
    });

    let ci = this.getCount() - 1;
    while (ci > 0) {
      const pi = (ci - 1) >> 1;
      if (this.queue[ci].priority > this.queue[pi].priority) {
        break;
      }
      const tmp = this.queue[ci];
      this.queue[ci] = this.queue[pi];
      this.queue[pi] = tmp;
      ci = pi;
    }
  }

  dequeue() {
    return this.queue.shift();
  }
}

export function getNeighbour(edges: Edge, node: string): string[] {
  for (let key in edges) {
    if (key === node) {
      return Object.keys(edges[key]);
    }
  }

  return [];
}

export function findAllPath(
  edges: Edge,
  originNode: string
): Record<string, string[]> {
  const frontier = new HeapPriorityQueue();
  frontier.enqueue(originNode, 0);

  const cameFrom: Record<string, string> = {
    [originNode]: "",
  };
  const costSoFar: Record<string, number> = {
    [originNode]: 0,
  };

  while (frontier.getCount() != 0) {
    const current: PriorityQueue = frontier.dequeue()!;
    const neighbours = getNeighbour(edges, current.item);
    neighbours.forEach((neighbour) => {
      const newCost = costSoFar[current.item] + edges[current.item][neighbour];
      if (newCost > 2) {
        return;
      }
      if (!costSoFar[neighbour] || newCost < costSoFar[neighbour]) {
        costSoFar[neighbour] = newCost;
        cameFrom[neighbour] = current.item;
        frontier.enqueue(neighbour, newCost);
      }
    });
  }

  const paths: Record<string, string[]> = {};
  Object.keys(cameFrom).forEach((destination) => {
    const path: string[] = [];
    let current = destination;
    while (current !== originNode) {
      path.push(current);
      current = cameFrom[current];
    }

    paths[destination] = path;
  });

  return paths;
}

File: lib/pathfinding.dart
import 'package:collection/collection.dart';

import "cell.dart";

typedef Edges = Map<Cell, Map<Cell, int>>;

class _CellItem {
  final Cell cell;
  final int priority;

  _CellItem(this.cell, this.priority);
}

class Pathfinding {
  final Edges edges;

  Pathfinding(this.edges);

  Map<Cell, List<Cell>> findAllPath(Cell originalNode, int movementPoint) {
    final playerNumber = originalNode.ship?.state.playerNumber ?? -1;
    final frontier = HeapPriorityQueue<_CellItem>((_CellItem itemA, _CellItem itemB) {
      return itemA.priority - itemB.priority;
    });
    frontier.add(_CellItem(originalNode, 0));

    final Map<Cell, Cell> cameFrom = {
      originalNode: originalNode,
    };
    final Map<Cell, int> costSoFar = {
      originalNode: 0,
    };

    while (frontier.isNotEmpty) {
      final current = frontier.removeFirst();
      final neighbours = _getNeighbours(current.cell);
      for (final neighbour in neighbours) {
        final newCost =
            costSoFar[current.cell]! + edges[current.cell]![neighbour]!;
        // Hostile ship or planet
        if (neighbour.ship != null &&
            neighbour.ship!.state.playerNumber != playerNumber) {
          continue;
        }
        if (neighbour.planet != null &&
            neighbour.planet!.attackable(playerNumber)) {
          continue;
        }
        if (newCost > movementPoint) {
          continue;
        }
        if (!costSoFar.containsKey(neighbour) ||
            (newCost < costSoFar[neighbour]!)) {
          costSoFar[neighbour] = newCost;
          cameFrom[neighbour] = current.cell;
          frontier.add(_CellItem(neighbour, newCost));
        }
      }
    }

    final Map<Cell, List<Cell>> paths = {};
    for (final destination in cameFrom.keys) {
      final List<Cell> path = [];
      Cell current = destination;
      // Cell is occupied by other ship
      if (current.ship != null) {
        continue;
      }
      while (current != originalNode) {
        path.add(current);
        current = cameFrom[current]!;
      }

      paths[destination] = path;
    }

    return paths;
  }

  Iterable<Cell> _getNeighbours(Cell node) {
    if (!edges.containsKey(node)) {
      return const [];
    }

    final map = edges[node]!;

    return map.keys;
  }
}
